File: apps/frontend/lib/workflow/evaluator/operators.ts
import type { ConditionOperator } from '@/lib/types/workflow'

const MS_PER_UNIT: Record<string, number> = {
  minutes: 60_000,
  hours: 3_600_000,
  days: 86_400_000,
  weeks: 604_800_000,
}

function applyCase(s: string, caseSensitive: boolean): string {
  return caseSensitive ? s : s.toLowerCase()
}

function asString(v: unknown): string | null {
  return typeof v === 'string' ? v : null
}

function asNumber(v: unknown): number | null {
  if (typeof v === 'number' && Number.isFinite(v)) return v
  return null
}

function asDateMs(v: unknown): number | null {
  if (v instanceof Date) return v.getTime()
  if (typeof v === 'string' && v) {
    const ms = Date.parse(v)
    return Number.isNaN(ms) ? null : ms
  }
  if (typeof v === 'number' && Number.isFinite(v)) return v
  return null
}

export interface OperatorContext {
  caseSensitive: boolean
}

export interface OperatorOutcome {
  matched: boolean
  actual: unknown
  expected: unknown
}

// Local-timezone semantics for date operators is intentional per spec §7 — Date.parse and
// new Date() both honor the runtime's local TZ.
export function applyOperator(
  operator: ConditionOperator,
  actual: unknown,
  expected: unknown,
  ctx: OperatorContext,
): OperatorOutcome {
  switch (operator) {
    case 'equals':
      return { matched: equalsOp(actual, expected, ctx), actual, expected }
    case 'contains':
    case 'starts_with':
    case 'ends_with':
      return { matched: stringOp(operator, actual, expected, ctx), actual, expected }
    case 'greater_than':
    case 'less_than':
    case 'greater_or_equal':
    case 'less_or_equal':
      return { matched: numberCompare(operator, actual, expected), actual, expected }
    case 'between':
      return { matched: betweenOp(actual, expected), actual, expected }
    case 'before':
    case 'after':
      return { matched: dateCompare(operator, actual, expected), actual, expected }
    case 'within_last':
      return { matched: withinLastOp(actual, expected), actual, expected }
  }
}

function equalsOp(actual: unknown, expected: unknown, ctx: OperatorContext): boolean {
  if (typeof actual === 'string' && typeof expected === 'string') {
    return applyCase(actual, ctx.caseSensitive) === applyCase(expected, ctx.caseSensitive)
  }
  return actual === expected
}

function stringOp(operator: 'contains' | 'starts_with' | 'ends_with', actual: unknown, expected: unknown, ctx: OperatorContext): boolean {
  const a = asString(actual)
  const e = asString(expected)
  if (a === null || e === null) return false
  const aa = applyCase(a, ctx.caseSensitive)
  const ee = applyCase(e, ctx.caseSensitive)
  if (operator === 'contains') return aa.includes(ee)
  if (operator === 'starts_with') return aa.startsWith(ee)
  return aa.endsWith(ee)
}

function numberCompare(operator: 'greater_than' | 'less_than' | 'greater_or_equal' | 'less_or_equal', actual: unknown, expected: unknown): boolean {
  const a = asNumber(actual)
  const e = asNumber(expected)
  if (a === null || e === null) return false
  if (operator === 'greater_than') return a > e
  if (operator === 'less_than') return a < e
  if (operator === 'greater_or_equal') return a >= e
  return a <= e
}

function betweenOp(actual: unknown, expected: unknown): boolean {
  if (!Array.isArray(expected) || expected.length !== 2) return false
  const [min, max] = expected
  const aNum = asNumber(actual)
  if (aNum !== null) {
    const minNum = asNumber(min)
    const maxNum = asNumber(max)
    if (minNum === null || maxNum === null) return false
    return aNum >= minNum && aNum <= maxNum
  }
  const aDate = asDateMs(actual)
  const minDate = asDateMs(min)
  const maxDate = asDateMs(max)
  if (aDate !== null && minDate !== null && maxDate !== null) {
    return aDate >= minDate && aDate <= maxDate
  }
  return false
}

function dateCompare(operator: 'before' | 'after', actual: unknown, expected: unknown): boolean {
  const a = asDateMs(actual)
  const e = asDateMs(expected)
  if (a === null || e === null) return false
  return operator === 'before' ? a < e : a > e
}

function withinLastOp(actual: unknown, expected: unknown): boolean {
  const a = asDateMs(actual)
  if (a === null) return false
  const window = parseDuration(expected)
  if (window === null) return false
  const now = Date.now()
  return a >= now - window && a <= now
}

function parseDuration(expected: unknown): number | null {
  if (!expected || typeof expected !== 'object') return null
  const obj = expected as { amount?: unknown; unit?: unknown }
  const amount = asNumber(obj.amount)
  const unit = typeof obj.unit === 'string' ? obj.unit : null
  if (amount === null || unit === null) return null
  const ms = MS_PER_UNIT[unit]
  if (ms === undefined) return null
  return amount * ms
}

File: apps/backend/app/modules/workflows/application/conditions/operators.py
"""Condition operators, mirroring the frontend evaluator (lib/workflow/evaluator/operators.ts).

Every operator is total: unusable operands evaluate to ``False`` rather than raising, so a
malformed condition can never abort a workflow run.
"""

import time
from datetime import datetime
from typing import Any, Optional

_MS_PER_UNIT = {
    "minutes": 60_000,
    "hours": 3_600_000,
    "days": 86_400_000,
    "weeks": 604_800_000,
}

_STRING_OPERATORS = frozenset({"contains", "starts_with", "ends_with"})
_NUMBER_OPERATORS = frozenset({"greater_than", "less_than", "greater_or_equal", "less_or_equal"})


def apply_operator(operator: str, actual: Any, expected: Any, case_sensitive: bool) -> bool:
    if operator == "equals":
        return _equals(actual, expected, case_sensitive)
    if operator in _STRING_OPERATORS:
        return _string_op(operator, actual, expected, case_sensitive)
    if operator in _NUMBER_OPERATORS:
        return _number_compare(operator, actual, expected)
    if operator == "between":
        return _between(actual, expected)
    if operator in ("before", "after"):
        return _date_compare(operator, actual, expected)
    if operator == "within_last":
        return _within_last(actual, expected)
    return False


def _is_number(value: Any) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool)


def _apply_case(value: str, case_sensitive: bool) -> str:
    return value if case_sensitive else value.lower()


def _as_string(value: Any) -> Optional[str]:
    return value if isinstance(value, str) else None


def _as_number(value: Any) -> Optional[float]:
    return float(value) if _is_number(value) else None


def _as_date_ms(value: Any) -> Optional[float]:
    if isinstance(value, datetime):
        return value.timestamp() * 1000
    if _is_number(value):
        return float(value)
    if isinstance(value, str) and value:
        return _parse_iso_ms(value)
    return None


def _parse_iso_ms(value: str) -> Optional[float]:
    candidate = value.replace("Z", "+00:00") if value.endswith("Z") else value
    try:
        parsed = datetime.fromisoformat(candidate)
    except ValueError:
        return None
    # Naive datetimes follow the runtime's local timezone, matching the frontend's
    # `new Date(string)` semantics for datetime-local inputs.
    return parsed.timestamp() * 1000


def _equals(actual: Any, expected: Any, case_sensitive: bool) -> bool:
    if isinstance(actual, str) and isinstance(expected, str):
        return _apply_case(actual, case_sensitive) == _apply_case(expected, case_sensitive)
    # Mirror JavaScript strict equality: differing types never match.
    if isinstance(actual, bool) or isinstance(expected, bool):
        return type(actual) is type(expected) and actual == expected
    if _is_number(actual) and _is_number(expected):
        return actual == expected
    return actual == expected


def _string_op(operator: str, actual: Any, expected: Any, case_sensitive: bool) -> bool:
    a = _as_string(actual)
    e = _as_string(expected)
    if a is None or e is None:
        return False
    a = _apply_case(a, case_sensitive)
    e = _apply_case(e, case_sensitive)
    if operator == "contains":
        return e in a
    if operator == "starts_with":
        return a.startswith(e)
    return a.endswith(e)


def _number_compare(operator: str, actual: Any, expected: Any) -> bool:
    a = _as_number(actual)
    e = _as_number(expected)
    if a is None or e is None:
        return False
    if operator == "greater_than":
        return a > e
    if operator == "less_than":
        return a < e
    if operator == "greater_or_equal":
        return a >= e
    return a <= e


def _between(actual: Any, expected: Any) -> bool:
    if not isinstance(expected, list) or len(expected) != 2:
        return False
    low, high = expected
    a_num = _as_number(actual)
    if a_num is not None:
        low_num = _as_number(low)
        high_num = _as_number(high)
        if low_num is None or high_num is None:
            return False
        return low_num <= a_num <= high_num
    a_date = _as_date_ms(actual)
    low_date = _as_date_ms(low)
    high_date = _as_date_ms(high)
    if a_date is not None and low_date is not None and high_date is not None:
        return low_date <= a_date <= high_date
    return False


def _date_compare(operator: str, actual: Any, expected: Any) -> bool:
    a = _as_date_ms(actual)
    e = _as_date_ms(expected)
    if a is None or e is None:
        return False
    return a < e if operator == "before" else a > e


def _within_last(actual: Any, expected: Any) -> bool:
    a = _as_date_ms(actual)
    if a is None:
        return False
    window = _parse_duration(expected)
    if window is None:
        return False
    now = time.time() * 1000
    return now - window <= a <= now


def _parse_duration(expected: Any) -> Optional[float]:
    if not isinstance(expected, dict):
        return None
    amount = _as_number(expected.get("amount"))
    unit = expected.get("unit")
    if amount is None or not isinstance(unit, str):
        return None
    ms = _MS_PER_UNIT.get(unit)
    return amount * ms if ms is not None else None
